File: src/0140-Word-Break-II/0140.cpp
#include <vector>
#include <string>
using namespace std;

static int x = []() {std::ios::sync_with_stdio(false); cin.tie(0); return 0; }();
class Solution 
{
public:
    vector<string> wordBreak(string s, vector<string>& wordDict) 
    {
        vector<string> out = _wordBreak(0, s, wordDict);
        for (int i = 0; i < out.size(); ++i) out[i].pop_back();
        return out;
    }

private:
    unordered_map<int, vector<string> > mem;
    
    vector<string> _wordBreak(int st, string& s, vector<string>& wordDict) 
    {
        if (mem.count(st)) return mem[st];

        for (int i = 0; i < wordDict.size(); ++i) 
        {
            if (s.substr(st, wordDict[i].length()) == wordDict[i]) 
            {
                vector<string> sublist = _wordBreak(st + wordDict[i].length(), s, wordDict);
                for(auto& it : sublist)
                  mem[st].push_back(wordDict[i] + " " +it);
            }
        }
        if (!mem.count(st) && st == s.length()) mem[st].push_back("");
        return mem[st];
    }
};

File: src/0140-Word-Break-II/0140.py
class Solution(object):
    def wordBreak(self, s, wordDict):
        """
        :type s: str
        :type wordDict: List[str]
        :rtype: List[str]
        """
        wordDict = set(wordDict)
        mem = dict()
        return self._wordBreak(s, wordDict, mem)
        
    def _wordBreak(self, s, wordDict, mem):
        if s in mem:
            return mem[s]
        
        res = list()
        if not s:
            res.append("")
            return res
        
        for i in range(len(s)+1):
            if s[:i] in wordDict:
                sublist = self._wordBreak(s[i:], wordDict, mem)
                for it in sublist:
                    res.append(str(s[:i]) + ('' if not it else ' ') + it)
                
        mem[s] = res
        return res            

if __name__ == "__main__":
    s = "catsanddog"
    wordDict = ["cat","cats","and","sand","dog"] 
    print(Solution().wordBreak(s, wordDict))